<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 7 Alpha diversity | Introduction to microbiome data science</title>
  <meta name="description" content="Course material" />
  <meta name="generator" content="bookdown 0.25 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 7 Alpha diversity | Introduction to microbiome data science" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="Course material" />
  <meta name="github-repo" content="microbiome/course_2021_radboud" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 7 Alpha diversity | Introduction to microbiome data science" />
  
  <meta name="twitter:description" content="Course material" />
  

<meta name="author" content="University of Turku" />


<meta name="date" content="2022-03-28" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="microbiome-data-exploration.html"/>
<link rel="next" href="beta-diversity.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Introduction to microbiome data science</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Overview<span></span></a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#introduction"><i class="fa fa-check"></i><b>1.1</b> Introduction<span></span></a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#learning-goals"><i class="fa fa-check"></i><b>1.2</b> Learning goals<span></span></a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#acknowledgments"><i class="fa fa-check"></i><b>1.3</b> Acknowledgments<span></span></a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="program.html"><a href="program.html"><i class="fa fa-check"></i><b>2</b> Program<span></span></a>
<ul>
<li class="chapter" data-level="2.1" data-path="program.html"><a href="program.html#monday-12-july-from-raw-sequences-to-ecological-data-analysis"><i class="fa fa-check"></i><b>2.1</b> Monday 12 July: from raw sequences to ecological data analysis<span></span></a></li>
<li class="chapter" data-level="2.2" data-path="program.html"><a href="program.html#tuesday-13-july---alpha-diversity"><i class="fa fa-check"></i><b>2.2</b> Tuesday 13 July - Alpha diversity<span></span></a></li>
<li class="chapter" data-level="2.3" data-path="program.html"><a href="program.html#wednesday-14-july---beta-diversity"><i class="fa fa-check"></i><b>2.3</b> Wednesday 14 July - Beta diversity<span></span></a></li>
<li class="chapter" data-level="2.4" data-path="program.html"><a href="program.html#thursday-15-july---differential-abundance"><i class="fa fa-check"></i><b>2.4</b> Thursday 15 July - Differential abundance<span></span></a></li>
<li class="chapter" data-level="2.5" data-path="program.html"><a href="program.html#friday-16-july-presentations-closing"><i class="fa fa-check"></i><b>2.5</b> Friday 16 July: Presentations &amp; closing<span></span></a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="getting-started.html"><a href="getting-started.html"><i class="fa fa-check"></i><b>3</b> Getting started<span></span></a>
<ul>
<li class="chapter" data-level="3.1" data-path="getting-started.html"><a href="getting-started.html#checklist-before-the-course"><i class="fa fa-check"></i><b>3.1</b> Checklist (before the course)<span></span></a></li>
<li class="chapter" data-level="3.2" data-path="getting-started.html"><a href="getting-started.html#support-and-resources"><i class="fa fa-check"></i><b>3.2</b> Support and resources<span></span></a></li>
<li class="chapter" data-level="3.3" data-path="getting-started.html"><a href="getting-started.html#installing-and-loading-the-required-r-packages"><i class="fa fa-check"></i><b>3.3</b> Installing and loading the required R packages<span></span></a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="reproducible-reporting-with-rmarkdown.html"><a href="reproducible-reporting-with-rmarkdown.html"><i class="fa fa-check"></i><b>4</b> Reproducible reporting with Rmarkdown<span></span></a></li>
<li class="chapter" data-level="5" data-path="importing-microbiome-data.html"><a href="importing-microbiome-data.html"><i class="fa fa-check"></i><b>5</b> Importing microbiome data<span></span></a>
<ul>
<li class="chapter" data-level="5.1" data-path="importing-microbiome-data.html"><a href="importing-microbiome-data.html#data-access"><i class="fa fa-check"></i><b>5.1</b> Data access<span></span></a></li>
<li class="chapter" data-level="5.2" data-path="importing-microbiome-data.html"><a href="importing-microbiome-data.html#importing-microbiome-data-in-r"><i class="fa fa-check"></i><b>5.2</b> Importing microbiome data in R<span></span></a></li>
<li class="chapter" data-level="5.3" data-path="importing-microbiome-data.html"><a href="importing-microbiome-data.html#example-solutions"><i class="fa fa-check"></i><b>5.3</b> Example solutions<span></span></a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="microbiome-data-exploration.html"><a href="microbiome-data-exploration.html"><i class="fa fa-check"></i><b>6</b> Microbiome data exploration<span></span></a>
<ul>
<li class="chapter" data-level="6.1" data-path="microbiome-data-exploration.html"><a href="microbiome-data-exploration.html#data-structure"><i class="fa fa-check"></i><b>6.1</b> Data structure<span></span></a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="microbiome-data-exploration.html"><a href="microbiome-data-exploration.html#transformations"><i class="fa fa-check"></i><b>6.1.1</b> Transformations<span></span></a></li>
<li class="chapter" data-level="6.1.2" data-path="microbiome-data-exploration.html"><a href="microbiome-data-exploration.html#aggregation"><i class="fa fa-check"></i><b>6.1.2</b> Aggregation<span></span></a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="microbiome-data-exploration.html"><a href="microbiome-data-exploration.html#visualization"><i class="fa fa-check"></i><b>6.2</b> Visualization<span></span></a></li>
<li class="chapter" data-level="6.3" data-path="microbiome-data-exploration.html"><a href="microbiome-data-exploration.html#exercises-optional"><i class="fa fa-check"></i><b>6.3</b> Exercises (optional)<span></span></a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="alpha-diversity.html"><a href="alpha-diversity.html"><i class="fa fa-check"></i><b>7</b> Alpha diversity<span></span></a>
<ul>
<li class="chapter" data-level="7.1" data-path="alpha-diversity.html"><a href="alpha-diversity.html#visualization-1"><i class="fa fa-check"></i><b>7.1</b> Visualization<span></span></a></li>
<li class="chapter" data-level="7.2" data-path="alpha-diversity.html"><a href="alpha-diversity.html#statistical-testing-and-comparisons"><i class="fa fa-check"></i><b>7.2</b> Statistical testing and comparisons<span></span></a></li>
<li class="chapter" data-level="7.3" data-path="alpha-diversity.html"><a href="alpha-diversity.html#exercises"><i class="fa fa-check"></i><b>7.3</b> Exercises<span></span></a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="beta-diversity.html"><a href="beta-diversity.html"><i class="fa fa-check"></i><b>8</b> Beta diversity<span></span></a>
<ul>
<li class="chapter" data-level="8.1" data-path="beta-diversity.html"><a href="beta-diversity.html#examples-of-pcoa-with-different-settings"><i class="fa fa-check"></i><b>8.1</b> Examples of PCoA with different settings<span></span></a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="beta-diversity.html"><a href="beta-diversity.html#pcoa-for-asv-level-data-with-bray-curtis"><i class="fa fa-check"></i><b>8.1.1</b> PCoA for ASV-level data with Bray-Curtis<span></span></a></li>
<li class="chapter" data-level="8.1.2" data-path="beta-diversity.html"><a href="beta-diversity.html#pcoa-for-asv-level-data-with-aitchison-distance"><i class="fa fa-check"></i><b>8.1.2</b> PCoA for ASV-level data with Aitchison distance<span></span></a></li>
<li class="chapter" data-level="8.1.3" data-path="beta-diversity.html"><a href="beta-diversity.html#pcoa-aggregated-to-phylum-level"><i class="fa fa-check"></i><b>8.1.3</b> PCoA aggregated to Phylum level<span></span></a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="beta-diversity.html"><a href="beta-diversity.html#highlighting-external-variables"><i class="fa fa-check"></i><b>8.2</b> Highlighting external variables<span></span></a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="beta-diversity.html"><a href="beta-diversity.html#discrete-grouping-variable-shown-with-colors"><i class="fa fa-check"></i><b>8.2.1</b> Discrete grouping variable shown with colors<span></span></a></li>
<li class="chapter" data-level="8.2.2" data-path="beta-diversity.html"><a href="beta-diversity.html#pcoa-plot-with-continuous-variable"><i class="fa fa-check"></i><b>8.2.2</b> PCoA plot with continuous variable<span></span></a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="beta-diversity.html"><a href="beta-diversity.html#estimating-associations-with-an-external-variable"><i class="fa fa-check"></i><b>8.3</b> Estimating associations with an external variable<span></span></a></li>
<li class="chapter" data-level="8.4" data-path="beta-diversity.html"><a href="beta-diversity.html#community-typing"><i class="fa fa-check"></i><b>8.4</b> Community typing<span></span></a></li>
<li class="chapter" data-level="8.5" data-path="beta-diversity.html"><a href="beta-diversity.html#exercises-1"><i class="fa fa-check"></i><b>8.5</b> Exercises<span></span></a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="differential-abundance-analysis.html"><a href="differential-abundance-analysis.html"><i class="fa fa-check"></i><b>9</b> Differential abundance analysis<span></span></a>
<ul>
<li class="chapter" data-level="9.1" data-path="differential-abundance-analysis.html"><a href="differential-abundance-analysis.html#wilcoxon-test"><i class="fa fa-check"></i><b>9.1</b> Wilcoxon test<span></span></a></li>
<li class="chapter" data-level="9.2" data-path="differential-abundance-analysis.html"><a href="differential-abundance-analysis.html#deseq2"><i class="fa fa-check"></i><b>9.2</b> DESeq2<span></span></a></li>
<li class="chapter" data-level="9.3" data-path="differential-abundance-analysis.html"><a href="differential-abundance-analysis.html#ancom-bc"><i class="fa fa-check"></i><b>9.3</b> ANCOM-BC<span></span></a></li>
<li class="chapter" data-level="9.4" data-path="differential-abundance-analysis.html"><a href="differential-abundance-analysis.html#comparison-of-the-methods"><i class="fa fa-check"></i><b>9.4</b> Comparison of the methods<span></span></a></li>
<li class="chapter" data-level="9.5" data-path="differential-abundance-analysis.html"><a href="differential-abundance-analysis.html#comparison-of-abundance"><i class="fa fa-check"></i><b>9.5</b> Comparison of abundance<span></span></a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="study-material.html"><a href="study-material.html"><i class="fa fa-check"></i><b>10</b> Study material<span></span></a>
<ul>
<li class="chapter" data-level="10.1" data-path="study-material.html"><a href="study-material.html#lecture-slides"><i class="fa fa-check"></i><b>10.1</b> Lecture slides<span></span></a></li>
<li class="chapter" data-level="10.2" data-path="study-material.html"><a href="study-material.html#r-programming-resources"><i class="fa fa-check"></i><b>10.2</b> R programming resources<span></span></a></li>
<li class="chapter" data-level="10.3" data-path="study-material.html"><a href="study-material.html#resources-for-treesummarizedexperiment"><i class="fa fa-check"></i><b>10.3</b> Resources for TreeSummarizedExperiment<span></span></a></li>
<li class="chapter" data-level="10.4" data-path="study-material.html"><a href="study-material.html#resources-for-phyloseq"><i class="fa fa-check"></i><b>10.4</b> Resources for phyloseq<span></span></a></li>
<li class="chapter" data-level="10.5" data-path="study-material.html"><a href="study-material.html#further-reading"><i class="fa fa-check"></i><b>10.5</b> Further reading<span></span></a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="miscellaneous-material.html"><a href="miscellaneous-material.html"><i class="fa fa-check"></i><b>11</b> Miscellaneous material<span></span></a>
<ul>
<li class="chapter" data-level="11.1" data-path="miscellaneous-material.html"><a href="miscellaneous-material.html#shapiro-wilk-test"><i class="fa fa-check"></i><b>11.1</b> Shapiro-Wilk test<span></span></a></li>
<li class="chapter" data-level="11.2" data-path="miscellaneous-material.html"><a href="miscellaneous-material.html#deseq-details"><i class="fa fa-check"></i><b>11.2</b> Deseq details<span></span></a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="exercise-solutions.html"><a href="exercise-solutions.html"><i class="fa fa-check"></i><b>12</b> Exercise Solutions<span></span></a>
<ul>
<li class="chapter" data-level="12.1" data-path="exercise-solutions.html"><a href="exercise-solutions.html#section-5"><i class="fa fa-check"></i><b>12.1</b> Section 5<span></span></a></li>
<li class="chapter" data-level="12.2" data-path="exercise-solutions.html"><a href="exercise-solutions.html#section-6"><i class="fa fa-check"></i><b>12.2</b> Section 6<span></span></a></li>
<li class="chapter" data-level="12.3" data-path="exercise-solutions.html"><a href="exercise-solutions.html#section-7"><i class="fa fa-check"></i><b>12.3</b> Section 7<span></span></a></li>
<li class="chapter" data-level="12.4" data-path="exercise-solutions.html"><a href="exercise-solutions.html#section-8"><i class="fa fa-check"></i><b>12.4</b> Section 8<span></span></a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Introduction to microbiome data science</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="alpha-diversity" class="section level1 hasAnchor" number="7">
<h1><span class="header-section-number">Chapter 7</span> Alpha diversity<a href="alpha-diversity.html#alpha-diversity" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>This section demonstrates the analysis of alpha diversity. This
quantity measures microbial diversity within each sample. Higher
numbers of unique taxa, and more even abundance distributions within a
sample yield larger values for alpha diversity.</p>
<p>Alpha diversity is a key quantity in a microbiome research. The <a href="https://microbiome.github.io/mia/"><em>mia</em>
package</a> provides access to a wide
variety of alpha diversity indices. For more background information
and examples with various alpha diversity indices, see the <a href="https://microbiome.github.io/OMA/microbiome-diversity.html#alpha-diversity">online
book</a>.</p>
<p>Let us show how to calculate to different diversity indices, Shannon
and Faith. Shannon index reflects how many different taxa there are
and how evenly they are distributed within a sample. Faith index
additionally takes into account the phylogenetic relations into
account.</p>
<div class="sourceCode" id="cb109"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb109-1"><a href="alpha-diversity.html#cb109-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Indices to be calculated. </span></span>
<span id="cb109-2"><a href="alpha-diversity.html#cb109-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Every index is calculated by default if we don&#39;t specify indices.</span></span>
<span id="cb109-3"><a href="alpha-diversity.html#cb109-3" aria-hidden="true" tabindex="-1"></a>indices <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;shannon&quot;</span>, <span class="st">&quot;faith&quot;</span>)</span>
<span id="cb109-4"><a href="alpha-diversity.html#cb109-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb109-5"><a href="alpha-diversity.html#cb109-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Indices are stored in colData (i.e., sample metadata). We can specify the name</span></span>
<span id="cb109-6"><a href="alpha-diversity.html#cb109-6" aria-hidden="true" tabindex="-1"></a><span class="co"># of column, or we can use the default name which is the name of index </span></span>
<span id="cb109-7"><a href="alpha-diversity.html#cb109-7" aria-hidden="true" tabindex="-1"></a><span class="co"># (i.e., &quot;shannon&quot; and &quot;faith&quot;). </span></span>
<span id="cb109-8"><a href="alpha-diversity.html#cb109-8" aria-hidden="true" tabindex="-1"></a>names <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;Shannon_diversity&quot;</span>, <span class="st">&quot;Faith_diversity&quot;</span>)</span>
<span id="cb109-9"><a href="alpha-diversity.html#cb109-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb109-10"><a href="alpha-diversity.html#cb109-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculates indices</span></span>
<span id="cb109-11"><a href="alpha-diversity.html#cb109-11" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">estimateDiversity</span>(tse, <span class="at">index =</span> indices, <span class="at">name =</span> names)</span>
<span id="cb109-12"><a href="alpha-diversity.html#cb109-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb109-13"><a href="alpha-diversity.html#cb109-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Shows the calculated indices</span></span>
<span id="cb109-14"><a href="alpha-diversity.html#cb109-14" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(<span class="fu">colData</span>(tse)[names])) <span class="sc">%&gt;%</span> </span>
<span id="cb109-15"><a href="alpha-diversity.html#cb109-15" aria-hidden="true" tabindex="-1"></a>  kableExtra<span class="sc">::</span><span class="fu">kable_styling</span>(<span class="st">&quot;striped&quot;</span>, </span>
<span id="cb109-16"><a href="alpha-diversity.html#cb109-16" aria-hidden="true" tabindex="-1"></a>                            <span class="at">latex_options=</span><span class="st">&quot;scale_down&quot;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb109-17"><a href="alpha-diversity.html#cb109-17" aria-hidden="true" tabindex="-1"></a>  kableExtra<span class="sc">::</span><span class="fu">scroll_box</span>(<span class="at">width =</span> <span class="st">&quot;100%&quot;</span>)</span></code></pre></div>
<div style="border: 1px solid #ddd; padding: 5px; overflow-x: scroll; width:100%; ">
<table class="table table-striped" style="margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
Shannon_diversity
</th>
<th style="text-align:right;">
Faith_diversity
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
A110
</td>
<td style="text-align:right;">
1.765407
</td>
<td style="text-align:right;">
7.39224
</td>
</tr>
<tr>
<td style="text-align:left;">
A12
</td>
<td style="text-align:right;">
2.716438
</td>
<td style="text-align:right;">
6.29378
</td>
</tr>
<tr>
<td style="text-align:left;">
A15
</td>
<td style="text-align:right;">
3.178103
</td>
<td style="text-align:right;">
6.60608
</td>
</tr>
<tr>
<td style="text-align:left;">
A19
</td>
<td style="text-align:right;">
2.891987
</td>
<td style="text-align:right;">
6.79708
</td>
</tr>
<tr>
<td style="text-align:left;">
A21
</td>
<td style="text-align:right;">
2.841979
</td>
<td style="text-align:right;">
6.65110
</td>
</tr>
<tr>
<td style="text-align:left;">
A23
</td>
<td style="text-align:right;">
2.797942
</td>
<td style="text-align:right;">
5.96246
</td>
</tr>
</tbody>
</table>
</div>
<p>Next we can visualize Shannon index with histogram.</p>
<div class="sourceCode" id="cb110"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb110-1"><a href="alpha-diversity.html#cb110-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ggplot needs data.frame format as input.</span></span>
<span id="cb110-2"><a href="alpha-diversity.html#cb110-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Here, colData is DataFrame, therefore it needs to be converted to data.frame</span></span>
<span id="cb110-3"><a href="alpha-diversity.html#cb110-3" aria-hidden="true" tabindex="-1"></a>shannon_hist <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="fu">as.data.frame</span>(<span class="fu">colData</span>(tse)), </span>
<span id="cb110-4"><a href="alpha-diversity.html#cb110-4" aria-hidden="true" tabindex="-1"></a>                       <span class="fu">aes</span>(<span class="at">x =</span> Shannon_diversity)) <span class="sc">+</span> </span>
<span id="cb110-5"><a href="alpha-diversity.html#cb110-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">20</span>, <span class="at">fill =</span> <span class="st">&quot;gray&quot;</span>, <span class="at">color =</span> <span class="st">&quot;black&quot;</span>) <span class="sc">+</span></span>
<span id="cb110-6"><a href="alpha-diversity.html#cb110-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Shannon index&quot;</span>, <span class="at">y =</span> <span class="st">&quot;Sample frequency&quot;</span>)</span>
<span id="cb110-7"><a href="alpha-diversity.html#cb110-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb110-8"><a href="alpha-diversity.html#cb110-8" aria-hidden="true" tabindex="-1"></a>shannon_hist</span></code></pre></div>
<p><img src="06-alpha_files/figure-html/unnamed-chunk-3-1.png" width="672" /></p>
<p>Next, let us compare the indices based on a scatter-plot.</p>
<div class="sourceCode" id="cb111"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb111-1"><a href="alpha-diversity.html#cb111-1" aria-hidden="true" tabindex="-1"></a>cross_plot <span class="ot">&lt;-</span> ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(<span class="fu">as.data.frame</span>(<span class="fu">colData</span>(tse)), </span>
<span id="cb111-2"><a href="alpha-diversity.html#cb111-2" aria-hidden="true" tabindex="-1"></a>                                     <span class="fu">aes</span>(<span class="at">x =</span> Shannon_diversity, <span class="at">y =</span> Faith_diversity)) <span class="sc">+</span> </span>
<span id="cb111-3"><a href="alpha-diversity.html#cb111-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="co"># Adds points</span></span>
<span id="cb111-4"><a href="alpha-diversity.html#cb111-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method=</span>lm) <span class="sc">+</span> <span class="co"># Adds regression line</span></span>
<span id="cb111-5"><a href="alpha-diversity.html#cb111-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Shannon index&quot;</span>, <span class="at">y =</span> <span class="st">&quot;Faith diversity&quot;</span>) </span>
<span id="cb111-6"><a href="alpha-diversity.html#cb111-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb111-7"><a href="alpha-diversity.html#cb111-7" aria-hidden="true" tabindex="-1"></a>cross_plot</span></code></pre></div>
<pre><code>## `geom_smooth()` using formula &#39;y ~ x&#39;</code></pre>
<p><img src="06-alpha_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
<div id="visualization-1" class="section level2 hasAnchor" number="7.1">
<h2><span class="header-section-number">7.1</span> Visualization<a href="alpha-diversity.html#visualization-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Next let us compare indices between different patient status and
cohorts. Boxplot is suitable for that purpose.</p>
<div class="sourceCode" id="cb113"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb113-1"><a href="alpha-diversity.html#cb113-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Creates Shannon boxplot </span></span>
<span id="cb113-2"><a href="alpha-diversity.html#cb113-2" aria-hidden="true" tabindex="-1"></a>shannon_box <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="fu">as.data.frame</span>(<span class="fu">colData</span>(tse)),</span>
<span id="cb113-3"><a href="alpha-diversity.html#cb113-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">aes</span>(<span class="at">x =</span> patient_status, </span>
<span id="cb113-4"><a href="alpha-diversity.html#cb113-4" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> Shannon_diversity,</span>
<span id="cb113-5"><a href="alpha-diversity.html#cb113-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">fill =</span> cohort)) <span class="sc">+</span> </span>
<span id="cb113-6"><a href="alpha-diversity.html#cb113-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb113-7"><a href="alpha-diversity.html#cb113-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">12</span>)) <span class="co"># makes titles smaller</span></span>
<span id="cb113-8"><a href="alpha-diversity.html#cb113-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb113-9"><a href="alpha-diversity.html#cb113-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Creates Faith boxplot </span></span>
<span id="cb113-10"><a href="alpha-diversity.html#cb113-10" aria-hidden="true" tabindex="-1"></a>faith_box <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="fu">as.data.frame</span>(<span class="fu">colData</span>(tse)), <span class="fu">aes</span>(<span class="at">x =</span> patient_status, </span>
<span id="cb113-11"><a href="alpha-diversity.html#cb113-11" aria-hidden="true" tabindex="-1"></a>                                                     <span class="at">y =</span> Faith_diversity, </span>
<span id="cb113-12"><a href="alpha-diversity.html#cb113-12" aria-hidden="true" tabindex="-1"></a>                                                     <span class="at">fill =</span> cohort)) <span class="sc">+</span> </span>
<span id="cb113-13"><a href="alpha-diversity.html#cb113-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb113-14"><a href="alpha-diversity.html#cb113-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">12</span>)) <span class="co"># makes titles smaller</span></span>
<span id="cb113-15"><a href="alpha-diversity.html#cb113-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb113-16"><a href="alpha-diversity.html#cb113-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Puts them into same picture</span></span>
<span id="cb113-17"><a href="alpha-diversity.html#cb113-17" aria-hidden="true" tabindex="-1"></a>gridExtra<span class="sc">::</span><span class="fu">grid.arrange</span>(shannon_box, faith_box, <span class="at">nrow =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="06-alpha_files/figure-html/unnamed-chunk-7-1.png" width="672" /></p>
<p>For an alternative visualization, see examples with <a href="https://microbiome.github.io/OMA/microbiome-diversity.html#alpha-diversity">scater::plotColData</a>.</p>
</div>
<div id="statistical-testing-and-comparisons" class="section level2 hasAnchor" number="7.2">
<h2><span class="header-section-number">7.2</span> Statistical testing and comparisons<a href="alpha-diversity.html#statistical-testing-and-comparisons" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>To further investigate if patient status could explain the variation
of Shannon index, let’s do a Wilcoxon test. This is a non-parametric
test that doesn’t make specific assumptions about the distribution,
unlike popular parametric tests, such as the t test, which assumes
normally distributed observations.</p>
<p>Wilcoxon test can be used to estimate whether the differences between
two groups is statistically significant. Here the ADHD and control
groups are not significantly different between groups (p-value is over
0.05).</p>
<div class="sourceCode" id="cb114"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb114-1"><a href="alpha-diversity.html#cb114-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Wilcoxon test, where Shannon index is the variable that we are comparing. </span></span>
<span id="cb114-2"><a href="alpha-diversity.html#cb114-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Patient status - ADHD or control - is the factor that we use for grouping. </span></span>
<span id="cb114-3"><a href="alpha-diversity.html#cb114-3" aria-hidden="true" tabindex="-1"></a>wilcoxon_shannon <span class="ot">&lt;-</span> <span class="fu">wilcox.test</span>(Shannon_diversity <span class="sc">~</span> patient_status, <span class="at">data =</span> <span class="fu">colData</span>(tse))</span>
<span id="cb114-4"><a href="alpha-diversity.html#cb114-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb114-5"><a href="alpha-diversity.html#cb114-5" aria-hidden="true" tabindex="-1"></a>wilcoxon_shannon</span></code></pre></div>
<pre><code>## 
##  Wilcoxon rank sum exact test
## 
## data:  Shannon_diversity by patient_status
## W = 76, p-value = 0.4879
## alternative hypothesis: true location shift is not equal to 0</code></pre>
<p>Another test that we can make is to test if ADHD samples differs between different
cohorts. From boxplot that we made in previous step, we can see that there might
be statistically significant difference between different cohorts.</p>
<p>Let’s compare Shannon index of ADHD samples between cohort 2 and cohort 3.</p>
<p>As we can see, there is statistically significant difference between the cohorts.</p>
<div class="sourceCode" id="cb116"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb116-1"><a href="alpha-diversity.html#cb116-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Takes subset of colData. Takes only ADHD samples</span></span>
<span id="cb116-2"><a href="alpha-diversity.html#cb116-2" aria-hidden="true" tabindex="-1"></a>ADHD_shannon <span class="ot">&lt;-</span> <span class="fu">colData</span>(tse)[ <span class="fu">colData</span>(tse)[, <span class="st">&quot;patient_status&quot;</span>] <span class="sc">==</span> <span class="st">&quot;ADHD&quot;</span> , ]</span>
<span id="cb116-3"><a href="alpha-diversity.html#cb116-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb116-4"><a href="alpha-diversity.html#cb116-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Takes subset of colData. Takes only samples that are in cohort 2 or cohort 3.</span></span>
<span id="cb116-5"><a href="alpha-diversity.html#cb116-5" aria-hidden="true" tabindex="-1"></a>ADHD_shannon <span class="ot">&lt;-</span> ADHD_shannon[ ADHD_shannon[, <span class="st">&quot;cohort&quot;</span>] <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;Cohort_2&quot;</span>, <span class="st">&quot;Cohort_3&quot;</span>) , ]</span>
<span id="cb116-6"><a href="alpha-diversity.html#cb116-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb116-7"><a href="alpha-diversity.html#cb116-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Wilcoxon test, where Shannon index is the variable that we are comparing. </span></span>
<span id="cb116-8"><a href="alpha-diversity.html#cb116-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Cohort - 2 or 3 - is the factor that we use for grouping. </span></span>
<span id="cb116-9"><a href="alpha-diversity.html#cb116-9" aria-hidden="true" tabindex="-1"></a>wilcoxon_shannon_ADHD_cohorts <span class="ot">&lt;-</span> <span class="fu">wilcox.test</span>(Shannon_diversity <span class="sc">~</span> cohort, <span class="at">data =</span> ADHD_shannon)</span>
<span id="cb116-10"><a href="alpha-diversity.html#cb116-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb116-11"><a href="alpha-diversity.html#cb116-11" aria-hidden="true" tabindex="-1"></a>wilcoxon_shannon_ADHD_cohorts</span></code></pre></div>
<pre><code>## 
##  Wilcoxon rank sum exact test
## 
## data:  Shannon_diversity by cohort
## W = 20, p-value = 0.01587
## alternative hypothesis: true location shift is not equal to 0</code></pre>
<p>For more examples, see a dedicated section on alpha diversity in the
<a href="https://microbiome.github.io/OMA/microbiome-diversity.html#alpha-diversity">online book</a>.</p>
</div>
<div id="exercises" class="section level2 hasAnchor" number="7.3">
<h2><span class="header-section-number">7.3</span> Exercises<a href="alpha-diversity.html#exercises" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Add the following in the reproducible summary report.</p>
<ul>
<li><p>Estimate alpha diversity for each sample and draw a histogram. Tip:
estimateDiversity</p></li>
<li><p>Compare the results between two or more
alpha diversity indices (visually and/or statistically).</p></li>
<li><p>See <a href="https://microbiome.github.io/OMA/microbiome-diversity.html#alpha-diversity">online book</a>
for further examples.</p></li>
<li><p>Example <a href="07-3-ex-sol-ADHD.html">Solutions</a></p></li>
</ul>

<pre><code>## Loading required package: SummarizedExperiment</code></pre>
<pre><code>## Loading required package: MatrixGenerics</code></pre>
<pre><code>## Loading required package: matrixStats</code></pre>
<pre><code>## 
## Attaching package: &#39;MatrixGenerics&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:matrixStats&#39;:
## 
##     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,
##     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,
##     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,
##     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,
##     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,
##     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,
##     colWeightedMeans, colWeightedMedians, colWeightedSds,
##     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,
##     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,
##     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,
##     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,
##     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,
##     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,
##     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,
##     rowWeightedSds, rowWeightedVars</code></pre>
<pre><code>## Loading required package: GenomicRanges</code></pre>
<pre><code>## Loading required package: stats4</code></pre>
<pre><code>## Loading required package: BiocGenerics</code></pre>
<pre><code>## 
## Attaching package: &#39;BiocGenerics&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:stats&#39;:
## 
##     IQR, mad, sd, var, xtabs</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     anyDuplicated, append, as.data.frame, basename, cbind, colnames,
##     dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,
##     grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,
##     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,
##     rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,
##     union, unique, unsplit, which.max, which.min</code></pre>
<pre><code>## Loading required package: S4Vectors</code></pre>
<pre><code>## 
## Attaching package: &#39;S4Vectors&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     expand.grid, I, unname</code></pre>
<pre><code>## Loading required package: IRanges</code></pre>
<pre><code>## Loading required package: GenomeInfoDb</code></pre>
<pre><code>## Loading required package: Biobase</code></pre>
<pre><code>## Welcome to Bioconductor
## 
##     Vignettes contain introductory material; view with
##     &#39;browseVignettes()&#39;. To cite Bioconductor, see
##     &#39;citation(&quot;Biobase&quot;)&#39;, and for packages &#39;citation(&quot;pkgname&quot;)&#39;.</code></pre>
<pre><code>## 
## Attaching package: &#39;Biobase&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:MatrixGenerics&#39;:
## 
##     rowMedians</code></pre>
<pre><code>## The following objects are masked from &#39;package:matrixStats&#39;:
## 
##     anyMissing, rowMedians</code></pre>
<pre><code>## Loading required package: SingleCellExperiment</code></pre>
<pre><code>## Loading required package: TreeSummarizedExperiment</code></pre>
<pre><code>## Loading required package: Biostrings</code></pre>
<pre><code>## Loading required package: XVector</code></pre>
<pre><code>## 
## Attaching package: &#39;Biostrings&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:base&#39;:
## 
##     strsplit</code></pre>
<pre><code>## Loading required package: MultiAssayExperiment</code></pre>
<pre><code>## Loading required package: ggplot2</code></pre>
<pre><code>## Loading required package: ggraph</code></pre>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="microbiome-data-exploration.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="beta-diversity.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["radboud2021_material.pdf"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
