<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 6 Microbiome data exploration | Introduction to microbiome data science</title>
  <meta name="description" content="Course material" />
  <meta name="generator" content="bookdown 0.26 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 6 Microbiome data exploration | Introduction to microbiome data science" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="Course material" />
  <meta name="github-repo" content="microbiome/course_2021_radboud" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 6 Microbiome data exploration | Introduction to microbiome data science" />
  
  <meta name="twitter:description" content="Course material" />
  

<meta name="author" content="University of Turku" />


<meta name="date" content="2022-05-17" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="importing-microbiome-data.html"/>
<link rel="next" href="alpha-diversity.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Introduction to microbiome data science</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Overview</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#introduction"><i class="fa fa-check"></i><b>1.1</b> Introduction</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#learning-goals"><i class="fa fa-check"></i><b>1.2</b> Learning goals</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#acknowledgments"><i class="fa fa-check"></i><b>1.3</b> Acknowledgments</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="program.html"><a href="program.html"><i class="fa fa-check"></i><b>2</b> Program</a>
<ul>
<li class="chapter" data-level="2.1" data-path="program.html"><a href="program.html#monday-12-july-from-raw-sequences-to-ecological-data-analysis"><i class="fa fa-check"></i><b>2.1</b> Monday 12 July: from raw sequences to ecological data analysis</a></li>
<li class="chapter" data-level="2.2" data-path="program.html"><a href="program.html#tuesday-13-july---alpha-diversity"><i class="fa fa-check"></i><b>2.2</b> Tuesday 13 July - Alpha diversity</a></li>
<li class="chapter" data-level="2.3" data-path="program.html"><a href="program.html#wednesday-14-july---beta-diversity"><i class="fa fa-check"></i><b>2.3</b> Wednesday 14 July - Beta diversity</a></li>
<li class="chapter" data-level="2.4" data-path="program.html"><a href="program.html#thursday-15-july---differential-abundance"><i class="fa fa-check"></i><b>2.4</b> Thursday 15 July - Differential abundance</a></li>
<li class="chapter" data-level="2.5" data-path="program.html"><a href="program.html#friday-16-july-presentations-closing"><i class="fa fa-check"></i><b>2.5</b> Friday 16 July: Presentations &amp; closing</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="getting-started.html"><a href="getting-started.html"><i class="fa fa-check"></i><b>3</b> Getting started</a>
<ul>
<li class="chapter" data-level="3.1" data-path="getting-started.html"><a href="getting-started.html#checklist-before-the-course"><i class="fa fa-check"></i><b>3.1</b> Checklist (before the course)</a></li>
<li class="chapter" data-level="3.2" data-path="getting-started.html"><a href="getting-started.html#support-and-resources"><i class="fa fa-check"></i><b>3.2</b> Support and resources</a></li>
<li class="chapter" data-level="3.3" data-path="getting-started.html"><a href="getting-started.html#installing-and-loading-the-required-r-packages"><i class="fa fa-check"></i><b>3.3</b> Installing and loading the required R packages</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="reproducible-reporting-with-rmarkdown.html"><a href="reproducible-reporting-with-rmarkdown.html"><i class="fa fa-check"></i><b>4</b> Reproducible reporting with Rmarkdown</a></li>
<li class="chapter" data-level="5" data-path="importing-microbiome-data.html"><a href="importing-microbiome-data.html"><i class="fa fa-check"></i><b>5</b> Importing microbiome data</a>
<ul>
<li class="chapter" data-level="5.1" data-path="importing-microbiome-data.html"><a href="importing-microbiome-data.html#data-access"><i class="fa fa-check"></i><b>5.1</b> Data access</a></li>
<li class="chapter" data-level="5.2" data-path="importing-microbiome-data.html"><a href="importing-microbiome-data.html#importing-microbiome-data-in-r"><i class="fa fa-check"></i><b>5.2</b> Importing microbiome data in R</a></li>
<li class="chapter" data-level="5.3" data-path="importing-microbiome-data.html"><a href="importing-microbiome-data.html#example-solutions"><i class="fa fa-check"></i><b>5.3</b> Example solutions</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="microbiome-data-exploration.html"><a href="microbiome-data-exploration.html"><i class="fa fa-check"></i><b>6</b> Microbiome data exploration</a>
<ul>
<li class="chapter" data-level="6.1" data-path="microbiome-data-exploration.html"><a href="microbiome-data-exploration.html#data-structure"><i class="fa fa-check"></i><b>6.1</b> Data structure</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="microbiome-data-exploration.html"><a href="microbiome-data-exploration.html#transformations"><i class="fa fa-check"></i><b>6.1.1</b> Transformations</a></li>
<li class="chapter" data-level="6.1.2" data-path="microbiome-data-exploration.html"><a href="microbiome-data-exploration.html#aggregation"><i class="fa fa-check"></i><b>6.1.2</b> Aggregation</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="microbiome-data-exploration.html"><a href="microbiome-data-exploration.html#visualization"><i class="fa fa-check"></i><b>6.2</b> Visualization</a></li>
<li class="chapter" data-level="6.3" data-path="microbiome-data-exploration.html"><a href="microbiome-data-exploration.html#exercises-optional"><i class="fa fa-check"></i><b>6.3</b> Exercises (optional)</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="alpha-diversity.html"><a href="alpha-diversity.html"><i class="fa fa-check"></i><b>7</b> Alpha diversity</a>
<ul>
<li class="chapter" data-level="7.1" data-path="alpha-diversity.html"><a href="alpha-diversity.html#visualization-1"><i class="fa fa-check"></i><b>7.1</b> Visualization</a></li>
<li class="chapter" data-level="7.2" data-path="alpha-diversity.html"><a href="alpha-diversity.html#statistical-testing-and-comparisons"><i class="fa fa-check"></i><b>7.2</b> Statistical testing and comparisons</a></li>
<li class="chapter" data-level="7.3" data-path="alpha-diversity.html"><a href="alpha-diversity.html#exercises"><i class="fa fa-check"></i><b>7.3</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="beta-diversity.html"><a href="beta-diversity.html"><i class="fa fa-check"></i><b>8</b> Beta diversity</a>
<ul>
<li class="chapter" data-level="8.1" data-path="beta-diversity.html"><a href="beta-diversity.html#examples-of-pcoa-with-different-settings"><i class="fa fa-check"></i><b>8.1</b> Examples of PCoA with different settings</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="beta-diversity.html"><a href="beta-diversity.html#pcoa-for-asv-level-data-with-bray-curtis"><i class="fa fa-check"></i><b>8.1.1</b> PCoA for ASV-level data with Bray-Curtis</a></li>
<li class="chapter" data-level="8.1.2" data-path="beta-diversity.html"><a href="beta-diversity.html#pcoa-for-asv-level-data-with-aitchison-distance"><i class="fa fa-check"></i><b>8.1.2</b> PCoA for ASV-level data with Aitchison distance</a></li>
<li class="chapter" data-level="8.1.3" data-path="beta-diversity.html"><a href="beta-diversity.html#pcoa-aggregated-to-phylum-level"><i class="fa fa-check"></i><b>8.1.3</b> PCoA aggregated to Phylum level</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="beta-diversity.html"><a href="beta-diversity.html#highlighting-external-variables"><i class="fa fa-check"></i><b>8.2</b> Highlighting external variables</a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="beta-diversity.html"><a href="beta-diversity.html#discrete-grouping-variable-shown-with-colors"><i class="fa fa-check"></i><b>8.2.1</b> Discrete grouping variable shown with colors</a></li>
<li class="chapter" data-level="8.2.2" data-path="beta-diversity.html"><a href="beta-diversity.html#pcoa-plot-with-continuous-variable"><i class="fa fa-check"></i><b>8.2.2</b> PCoA plot with continuous variable</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="beta-diversity.html"><a href="beta-diversity.html#estimating-associations-with-an-external-variable"><i class="fa fa-check"></i><b>8.3</b> Estimating associations with an external variable</a></li>
<li class="chapter" data-level="8.4" data-path="beta-diversity.html"><a href="beta-diversity.html#community-typing"><i class="fa fa-check"></i><b>8.4</b> Community typing</a></li>
<li class="chapter" data-level="8.5" data-path="beta-diversity.html"><a href="beta-diversity.html#exercises-1"><i class="fa fa-check"></i><b>8.5</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="differential-abundance-analysis.html"><a href="differential-abundance-analysis.html"><i class="fa fa-check"></i><b>9</b> Differential abundance analysis</a>
<ul>
<li class="chapter" data-level="9.1" data-path="differential-abundance-analysis.html"><a href="differential-abundance-analysis.html#wilcoxon-test"><i class="fa fa-check"></i><b>9.1</b> Wilcoxon test</a></li>
<li class="chapter" data-level="9.2" data-path="differential-abundance-analysis.html"><a href="differential-abundance-analysis.html#deseq2"><i class="fa fa-check"></i><b>9.2</b> DESeq2</a></li>
<li class="chapter" data-level="9.3" data-path="differential-abundance-analysis.html"><a href="differential-abundance-analysis.html#ancom-bc"><i class="fa fa-check"></i><b>9.3</b> ANCOM-BC</a></li>
<li class="chapter" data-level="9.4" data-path="differential-abundance-analysis.html"><a href="differential-abundance-analysis.html#comparison-of-the-methods"><i class="fa fa-check"></i><b>9.4</b> Comparison of the methods</a></li>
<li class="chapter" data-level="9.5" data-path="differential-abundance-analysis.html"><a href="differential-abundance-analysis.html#comparison-of-abundance"><i class="fa fa-check"></i><b>9.5</b> Comparison of abundance</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="study-material.html"><a href="study-material.html"><i class="fa fa-check"></i><b>10</b> Study material</a>
<ul>
<li class="chapter" data-level="10.1" data-path="study-material.html"><a href="study-material.html#lecture-slides"><i class="fa fa-check"></i><b>10.1</b> Lecture slides</a></li>
<li class="chapter" data-level="10.2" data-path="study-material.html"><a href="study-material.html#r-programming-resources"><i class="fa fa-check"></i><b>10.2</b> R programming resources</a></li>
<li class="chapter" data-level="10.3" data-path="study-material.html"><a href="study-material.html#resources-for-treesummarizedexperiment"><i class="fa fa-check"></i><b>10.3</b> Resources for TreeSummarizedExperiment</a></li>
<li class="chapter" data-level="10.4" data-path="study-material.html"><a href="study-material.html#resources-for-phyloseq"><i class="fa fa-check"></i><b>10.4</b> Resources for phyloseq</a></li>
<li class="chapter" data-level="10.5" data-path="study-material.html"><a href="study-material.html#further-reading"><i class="fa fa-check"></i><b>10.5</b> Further reading</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="miscellaneous-material.html"><a href="miscellaneous-material.html"><i class="fa fa-check"></i><b>11</b> Miscellaneous material</a>
<ul>
<li class="chapter" data-level="11.1" data-path="miscellaneous-material.html"><a href="miscellaneous-material.html#shapiro-wilk-test"><i class="fa fa-check"></i><b>11.1</b> Shapiro-Wilk test</a></li>
<li class="chapter" data-level="11.2" data-path="miscellaneous-material.html"><a href="miscellaneous-material.html#deseq-details"><i class="fa fa-check"></i><b>11.2</b> Deseq details</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="exercise-solutions.html"><a href="exercise-solutions.html"><i class="fa fa-check"></i><b>12</b> Exercise Solutions</a>
<ul>
<li class="chapter" data-level="12.1" data-path="exercise-solutions.html"><a href="exercise-solutions.html#section-5"><i class="fa fa-check"></i><b>12.1</b> Section 5</a></li>
<li class="chapter" data-level="12.2" data-path="exercise-solutions.html"><a href="exercise-solutions.html#section-6"><i class="fa fa-check"></i><b>12.2</b> Section 6</a></li>
<li class="chapter" data-level="12.3" data-path="exercise-solutions.html"><a href="exercise-solutions.html#section-7"><i class="fa fa-check"></i><b>12.3</b> Section 7</a></li>
<li class="chapter" data-level="12.4" data-path="exercise-solutions.html"><a href="exercise-solutions.html#section-8"><i class="fa fa-check"></i><b>12.4</b> Section 8</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Introduction to microbiome data science</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="microbiome-data-exploration" class="section level1 hasAnchor" number="6">
<h1><span class="header-section-number">Chapter 6</span> Microbiome data exploration<a href="microbiome-data-exploration.html#microbiome-data-exploration" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Now we have loaded the data set into R. Next, let us walk through some
basic operations for data exploration to confirm that the data has all
the necessary components.</p>
<div id="data-structure" class="section level2 hasAnchor" number="6.1">
<h2><span class="header-section-number">6.1</span> Data structure<a href="microbiome-data-exploration.html#data-structure" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Let us now investigate how taxonomic profiling data is organized in R.</p>
<p>Dimensionality tells us how many taxa and samples the data
contains. As we can see, there are 151 taxa and 27
samples.</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="microbiome-data-exploration.html#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(tse)</span></code></pre></div>
<pre><code>## [1] 151  27</code></pre>
<p>The <code>rowData</code> slot contains a taxonomic table. This includes taxonomic
information for each of the 151 entries. With the <code>head()</code>
command, we can print just the beginning of the table.</p>
<p>The <code>rowData</code> seems to contain information from 6
different taxonomy classes.</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="microbiome-data-exploration.html#cb50-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(<span class="fu">rowData</span>(tse))) <span class="sc">%&gt;%</span> </span>
<span id="cb50-2"><a href="microbiome-data-exploration.html#cb50-2" aria-hidden="true" tabindex="-1"></a>  kableExtra<span class="sc">::</span><span class="fu">kable_styling</span>(<span class="st">&quot;striped&quot;</span>, </span>
<span id="cb50-3"><a href="microbiome-data-exploration.html#cb50-3" aria-hidden="true" tabindex="-1"></a>                            <span class="at">latex_options=</span><span class="st">&quot;scale_down&quot;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb50-4"><a href="microbiome-data-exploration.html#cb50-4" aria-hidden="true" tabindex="-1"></a>  kableExtra<span class="sc">::</span><span class="fu">scroll_box</span>(<span class="at">width =</span> <span class="st">&quot;100%&quot;</span>)</span></code></pre></div>
<div style="border: 1px solid #ddd; padding: 5px; overflow-x: scroll; width:100%; ">
<table class="table table-striped" style="margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
Kingdom
</th>
<th style="text-align:left;">
Phylum
</th>
<th style="text-align:left;">
Class
</th>
<th style="text-align:left;">
Order
</th>
<th style="text-align:left;">
Family
</th>
<th style="text-align:left;">
Genus
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
1726470
</td>
<td style="text-align:left;">
Bacteria
</td>
<td style="text-align:left;">
Bacteroidetes
</td>
<td style="text-align:left;">
Bacteroidia
</td>
<td style="text-align:left;">
Bacteroidales
</td>
<td style="text-align:left;">
Bacteroidaceae
</td>
<td style="text-align:left;">
Bacteroides
</td>
</tr>
<tr>
<td style="text-align:left;">
1726471
</td>
<td style="text-align:left;">
Bacteria
</td>
<td style="text-align:left;">
Bacteroidetes
</td>
<td style="text-align:left;">
Bacteroidia
</td>
<td style="text-align:left;">
Bacteroidales
</td>
<td style="text-align:left;">
Bacteroidaceae
</td>
<td style="text-align:left;">
Bacteroides
</td>
</tr>
<tr>
<td style="text-align:left;">
17264731
</td>
<td style="text-align:left;">
Bacteria
</td>
<td style="text-align:left;">
Bacteroidetes
</td>
<td style="text-align:left;">
Bacteroidia
</td>
<td style="text-align:left;">
Bacteroidales
</td>
<td style="text-align:left;">
Porphyromonadaceae
</td>
<td style="text-align:left;">
Parabacteroides
</td>
</tr>
<tr>
<td style="text-align:left;">
17264726
</td>
<td style="text-align:left;">
Bacteria
</td>
<td style="text-align:left;">
Bacteroidetes
</td>
<td style="text-align:left;">
Bacteroidia
</td>
<td style="text-align:left;">
Bacteroidales
</td>
<td style="text-align:left;">
Bacteroidaceae
</td>
<td style="text-align:left;">
Bacteroides
</td>
</tr>
<tr>
<td style="text-align:left;">
1726472
</td>
<td style="text-align:left;">
Bacteria
</td>
<td style="text-align:left;">
Verrucomicrobia
</td>
<td style="text-align:left;">
Verrucomicrobiae
</td>
<td style="text-align:left;">
Verrucomicrobiales
</td>
<td style="text-align:left;">
Verrucomicrobiaceae
</td>
<td style="text-align:left;">
Akkermansia
</td>
</tr>
<tr>
<td style="text-align:left;">
17264724
</td>
<td style="text-align:left;">
Bacteria
</td>
<td style="text-align:left;">
Bacteroidetes
</td>
<td style="text-align:left;">
Bacteroidia
</td>
<td style="text-align:left;">
Bacteroidales
</td>
<td style="text-align:left;">
Bacteroidaceae
</td>
<td style="text-align:left;">
Bacteroides
</td>
</tr>
</tbody>
</table>
</div>
<p>The colData slot contains sample metadata. It contains information for all 27 samples.
However, here only the 6 first samples are shown as we use the <code>head()</code> command. There
are 4 columns, that contain information, e.g., about patients’ status, and cohort.</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="microbiome-data-exploration.html#cb51-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(<span class="fu">colData</span>(tse))) <span class="sc">%&gt;%</span> </span>
<span id="cb51-2"><a href="microbiome-data-exploration.html#cb51-2" aria-hidden="true" tabindex="-1"></a>  kableExtra<span class="sc">::</span><span class="fu">kable_styling</span>(<span class="st">&quot;striped&quot;</span>, </span>
<span id="cb51-3"><a href="microbiome-data-exploration.html#cb51-3" aria-hidden="true" tabindex="-1"></a>                            <span class="at">latex_options=</span><span class="st">&quot;scale_down&quot;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb51-4"><a href="microbiome-data-exploration.html#cb51-4" aria-hidden="true" tabindex="-1"></a>  kableExtra<span class="sc">::</span><span class="fu">scroll_box</span>(<span class="at">width =</span> <span class="st">&quot;100%&quot;</span>)</span></code></pre></div>
<div style="border: 1px solid #ddd; padding: 5px; overflow-x: scroll; width:100%; ">
<table class="table table-striped" style="margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
patient_status
</th>
<th style="text-align:left;">
cohort
</th>
<th style="text-align:left;">
patient_status_vs_cohort
</th>
<th style="text-align:left;">
sample_name
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
A110
</td>
<td style="text-align:left;">
ADHD
</td>
<td style="text-align:left;">
Cohort_1
</td>
<td style="text-align:left;">
ADHD_Cohort_1
</td>
<td style="text-align:left;">
A110
</td>
</tr>
<tr>
<td style="text-align:left;">
A12
</td>
<td style="text-align:left;">
ADHD
</td>
<td style="text-align:left;">
Cohort_1
</td>
<td style="text-align:left;">
ADHD_Cohort_1
</td>
<td style="text-align:left;">
A12
</td>
</tr>
<tr>
<td style="text-align:left;">
A15
</td>
<td style="text-align:left;">
ADHD
</td>
<td style="text-align:left;">
Cohort_1
</td>
<td style="text-align:left;">
ADHD_Cohort_1
</td>
<td style="text-align:left;">
A15
</td>
</tr>
<tr>
<td style="text-align:left;">
A19
</td>
<td style="text-align:left;">
ADHD
</td>
<td style="text-align:left;">
Cohort_1
</td>
<td style="text-align:left;">
ADHD_Cohort_1
</td>
<td style="text-align:left;">
A19
</td>
</tr>
<tr>
<td style="text-align:left;">
A21
</td>
<td style="text-align:left;">
ADHD
</td>
<td style="text-align:left;">
Cohort_2
</td>
<td style="text-align:left;">
ADHD_Cohort_2
</td>
<td style="text-align:left;">
A21
</td>
</tr>
<tr>
<td style="text-align:left;">
A23
</td>
<td style="text-align:left;">
ADHD
</td>
<td style="text-align:left;">
Cohort_2
</td>
<td style="text-align:left;">
ADHD_Cohort_2
</td>
<td style="text-align:left;">
A23
</td>
</tr>
</tbody>
</table>
</div>
<p>From here, we can draw summaries of the sample (column) data, for
instance to see what is the patient status distribution.</p>
<p>The command <code>colData(tse)$patient_status</code> fetches the data from the
column, and <code>table()</code> creates a table that shows how many times each
class is present, and <code>sort()</code> sorts the table to ascending order.</p>
<p>There are 13
samples from patients having ADHD,
and 14 control samples.</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="microbiome-data-exploration.html#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sort</span>(<span class="fu">table</span>(<span class="fu">colData</span>(tse)<span class="sc">$</span>patient_status))</span></code></pre></div>
<pre><code>## 
##    ADHD Control 
##      13      14</code></pre>
<div id="transformations" class="section level3 hasAnchor" number="6.1.1">
<h3><span class="header-section-number">6.1.1</span> Transformations<a href="microbiome-data-exploration.html#transformations" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Microbial abundances are typically ‘compositional’ (relative) in the
current microbiome profiling data sets. This is due to technical
aspects of the data generation process (see e.g. <a href="https://www.frontiersin.org/articles/10.3389/fmicb.2017.02224/full">Gloor et al.,
2017</a>).</p>
<p>The next example calculates relative abundances as these are usually easier to
interpret than plain counts. For some statistical models we need to
transform the data into other formats as explained in above link (and
as we will see later).</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="microbiome-data-exploration.html#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculates relative abundances, and stores the table to assays</span></span>
<span id="cb54-2"><a href="microbiome-data-exploration.html#cb54-2" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">transformCounts</span>(tse, <span class="at">method =</span> <span class="st">&quot;relabundance&quot;</span>)</span></code></pre></div>
<p>A variety of standard transformations for microbiome data are available for <code>TSE</code> data objects through <a href="https://microbiome.github.io/mia/reference/transformCounts.html">mia R package</a>.</p>
</div>
<div id="aggregation" class="section level3 hasAnchor" number="6.1.2">
<h3><span class="header-section-number">6.1.2</span> Aggregation<a href="microbiome-data-exploration.html#aggregation" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Microbial species can be called at multiple taxonomic resolutions. We
can easily agglomerate the data based on taxonomic ranks. Here, we
agglomerate the data at Phylum level.</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb55-1"><a href="microbiome-data-exploration.html#cb55-1" aria-hidden="true" tabindex="-1"></a>tse_phylum <span class="ot">&lt;-</span> <span class="fu">agglomerateByRank</span>(tse, <span class="at">rank =</span> <span class="st">&quot;Phylum&quot;</span>)</span>
<span id="cb55-2"><a href="microbiome-data-exploration.html#cb55-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-3"><a href="microbiome-data-exploration.html#cb55-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Show dimensionality</span></span>
<span id="cb55-4"><a href="microbiome-data-exploration.html#cb55-4" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(tse_phylum)</span></code></pre></div>
<pre><code>## [1]  5 27</code></pre>
<p>Now there are 5 taxa and 27
samples, meaning that there are 5 different
Phylum level taxonomic groups. Looking at the <code>rowData</code> after
agglomeration shows all Firmicutes are combined together, and all
lower rank information is lost.</p>
<p>From the assay we can see that all abundances of taxa that belong to
Firmicutes are summed up.</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="microbiome-data-exploration.html#cb57-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(<span class="fu">rowData</span>(tse_phylum))) <span class="sc">%&gt;%</span> </span>
<span id="cb57-2"><a href="microbiome-data-exploration.html#cb57-2" aria-hidden="true" tabindex="-1"></a>  kableExtra<span class="sc">::</span><span class="fu">kable_styling</span>(<span class="st">&quot;striped&quot;</span>, </span>
<span id="cb57-3"><a href="microbiome-data-exploration.html#cb57-3" aria-hidden="true" tabindex="-1"></a>                            <span class="at">latex_options=</span><span class="st">&quot;scale_down&quot;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb57-4"><a href="microbiome-data-exploration.html#cb57-4" aria-hidden="true" tabindex="-1"></a>  kableExtra<span class="sc">::</span><span class="fu">scroll_box</span>(<span class="at">width =</span> <span class="st">&quot;100%&quot;</span>)</span></code></pre></div>
<div style="border: 1px solid #ddd; padding: 5px; overflow-x: scroll; width:100%; ">
<table class="table table-striped" style="margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
Kingdom
</th>
<th style="text-align:left;">
Phylum
</th>
<th style="text-align:left;">
Class
</th>
<th style="text-align:left;">
Order
</th>
<th style="text-align:left;">
Family
</th>
<th style="text-align:left;">
Genus
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Bacteroidetes
</td>
<td style="text-align:left;">
Bacteria
</td>
<td style="text-align:left;">
Bacteroidetes
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
</tr>
<tr>
<td style="text-align:left;">
Verrucomicrobia
</td>
<td style="text-align:left;">
Bacteria
</td>
<td style="text-align:left;">
Verrucomicrobia
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
</tr>
<tr>
<td style="text-align:left;">
Proteobacteria
</td>
<td style="text-align:left;">
Bacteria
</td>
<td style="text-align:left;">
Proteobacteria
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
</tr>
<tr>
<td style="text-align:left;">
Firmicutes
</td>
<td style="text-align:left;">
Bacteria
</td>
<td style="text-align:left;">
Firmicutes
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
</tr>
<tr>
<td style="text-align:left;">
Cyanobacteria
</td>
<td style="text-align:left;">
Bacteria
</td>
<td style="text-align:left;">
Cyanobacteria
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
</tr>
</tbody>
</table>
</div>
<p>If you are sharp, you have by now noticed that all the aggregated
values in the above example are NA’s (missing data). This is because
the agglomeration is missing abundances for certain taxa, and in that
case the sum is not defined by default (<code>na.rm = FALSE</code>). We can
ignore the missing values in summing up the data by setting <code>na.rm = TRUE</code>; then the taxa that do not have information in specified level
will be removed. Those taxa that do not have information in specified
level are agglomerated at lowest possible level that is left after
agglomeration.</p>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="microbiome-data-exploration.html#cb58-1" aria-hidden="true" tabindex="-1"></a>temp <span class="ot">&lt;-</span> <span class="fu">rowData</span>(<span class="fu">agglomerateByRank</span>(tse, <span class="at">rank =</span> <span class="st">&quot;Genus&quot;</span>))</span>
<span id="cb58-2"><a href="microbiome-data-exploration.html#cb58-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-3"><a href="microbiome-data-exploration.html#cb58-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Prints those taxa that do not have information at the Genus level</span></span>
<span id="cb58-4"><a href="microbiome-data-exploration.html#cb58-4" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(temp[temp<span class="sc">$</span>Genus <span class="sc">==</span> <span class="st">&quot;&quot;</span>,])) <span class="sc">%&gt;%</span> </span>
<span id="cb58-5"><a href="microbiome-data-exploration.html#cb58-5" aria-hidden="true" tabindex="-1"></a>  kableExtra<span class="sc">::</span><span class="fu">kable_styling</span>(<span class="st">&quot;striped&quot;</span>, </span>
<span id="cb58-6"><a href="microbiome-data-exploration.html#cb58-6" aria-hidden="true" tabindex="-1"></a>                            <span class="at">latex_options=</span><span class="st">&quot;scale_down&quot;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb58-7"><a href="microbiome-data-exploration.html#cb58-7" aria-hidden="true" tabindex="-1"></a>  kableExtra<span class="sc">::</span><span class="fu">scroll_box</span>(<span class="at">width =</span> <span class="st">&quot;100%&quot;</span>)</span></code></pre></div>
<div style="border: 1px solid #ddd; padding: 5px; overflow-x: scroll; width:100%; ">
<table class="table table-striped" style="margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
Kingdom
</th>
<th style="text-align:left;">
Phylum
</th>
<th style="text-align:left;">
Class
</th>
<th style="text-align:left;">
Order
</th>
<th style="text-align:left;">
Family
</th>
<th style="text-align:left;">
Genus
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Family:Lachnospiraceae
</td>
<td style="text-align:left;">
Bacteria
</td>
<td style="text-align:left;">
Firmicutes
</td>
<td style="text-align:left;">
Clostridia
</td>
<td style="text-align:left;">
Clostridiales
</td>
<td style="text-align:left;">
Lachnospiraceae
</td>
<td style="text-align:left;">
</td>
</tr>
<tr>
<td style="text-align:left;">
Order:Bacteroidales
</td>
<td style="text-align:left;">
Bacteria
</td>
<td style="text-align:left;">
Bacteroidetes
</td>
<td style="text-align:left;">
Bacteroidia
</td>
<td style="text-align:left;">
Bacteroidales
</td>
<td style="text-align:left;">
</td>
<td style="text-align:left;">
</td>
</tr>
<tr>
<td style="text-align:left;">
Order:Clostridiales
</td>
<td style="text-align:left;">
Bacteria
</td>
<td style="text-align:left;">
Firmicutes
</td>
<td style="text-align:left;">
Clostridia
</td>
<td style="text-align:left;">
Clostridiales
</td>
<td style="text-align:left;">
</td>
<td style="text-align:left;">
</td>
</tr>
<tr>
<td style="text-align:left;">
Family:Enterobacteriaceae
</td>
<td style="text-align:left;">
Bacteria
</td>
<td style="text-align:left;">
Proteobacteria
</td>
<td style="text-align:left;">
Gammaproteobacteria
</td>
<td style="text-align:left;">
Enterobacteriales
</td>
<td style="text-align:left;">
Enterobacteriaceae
</td>
<td style="text-align:left;">
</td>
</tr>
<tr>
<td style="text-align:left;">
Order:Gastranaerophilales
</td>
<td style="text-align:left;">
Bacteria
</td>
<td style="text-align:left;">
Cyanobacteria
</td>
<td style="text-align:left;">
Melainabacteria
</td>
<td style="text-align:left;">
Gastranaerophilales
</td>
<td style="text-align:left;">
</td>
<td style="text-align:left;">
</td>
</tr>
</tbody>
</table>
</div>
<p>Here agglomeration is done similarly, but na.rm = TRUE</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb59-1"><a href="microbiome-data-exploration.html#cb59-1" aria-hidden="true" tabindex="-1"></a>temp2 <span class="ot">&lt;-</span> <span class="fu">rowData</span>(<span class="fu">agglomerateByRank</span>(tse, <span class="at">rank =</span> <span class="st">&quot;Genus&quot;</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>))</span>
<span id="cb59-2"><a href="microbiome-data-exploration.html#cb59-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-3"><a href="microbiome-data-exploration.html#cb59-3" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">&quot;Agglomeration with na.rm = FALSE: &quot;</span>, <span class="fu">dim</span>(temp)[<span class="dv">1</span>], <span class="st">&quot; taxa.&quot;</span>))</span></code></pre></div>
<pre><code>## [1] &quot;Agglomeration with na.rm = FALSE: 54 taxa.&quot;</code></pre>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb61-1"><a href="microbiome-data-exploration.html#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">&quot;Agglomeration with na.rm = TRUE: &quot;</span>, <span class="fu">dim</span>(temp2)[<span class="dv">1</span>], <span class="st">&quot; taxa.&quot;</span>))</span></code></pre></div>
<pre><code>## [1] &quot;Agglomeration with na.rm = TRUE: 49 taxa.&quot;</code></pre>
<p>The <a href="https://microbiome.github.io/mia/reference/index.html">mia
package</a>
contains further examples on various data agglomeration and splitting
options.</p>
</div>
</div>
<div id="visualization" class="section level2 hasAnchor" number="6.2">
<h2><span class="header-section-number">6.2</span> Visualization<a href="microbiome-data-exploration.html#visualization" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The <a href="https://microbiome.github.io/miaViz/">miaViz package</a> facilitates
data visualization. Let us plot the Phylum level abundances.</p>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb63-1"><a href="microbiome-data-exploration.html#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Here we specify &quot;relabundance&quot; to be abundance table that we use for plotting.</span></span>
<span id="cb63-2"><a href="microbiome-data-exploration.html#cb63-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Note that we can use agglomerated or non-agglomerated tse as an input, because</span></span>
<span id="cb63-3"><a href="microbiome-data-exploration.html#cb63-3" aria-hidden="true" tabindex="-1"></a><span class="co"># the function agglomeration is built-in option. </span></span>
<span id="cb63-4"><a href="microbiome-data-exploration.html#cb63-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-5"><a href="microbiome-data-exploration.html#cb63-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Legend does not fit into picture, so its height is reduced.</span></span>
<span id="cb63-6"><a href="microbiome-data-exploration.html#cb63-6" aria-hidden="true" tabindex="-1"></a>plot_abundance <span class="ot">&lt;-</span> <span class="fu">plotAbundance</span>(tse, <span class="at">abund_values=</span><span class="st">&quot;relabundance&quot;</span>, <span class="at">rank =</span> <span class="st">&quot;Phylum&quot;</span>) <span class="sc">+</span></span>
<span id="cb63-7"><a href="microbiome-data-exploration.html#cb63-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.key.height =</span> <span class="fu">unit</span>(<span class="fl">0.5</span>, <span class="st">&quot;cm&quot;</span>)) <span class="sc">+</span></span>
<span id="cb63-8"><a href="microbiome-data-exploration.html#cb63-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">label =</span> scales<span class="sc">::</span>percent)</span></code></pre></div>
<pre><code>## Scale for &#39;y&#39; is already present. Adding another scale for &#39;y&#39;, which will
## replace the existing scale.</code></pre>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb65-1"><a href="microbiome-data-exploration.html#cb65-1" aria-hidden="true" tabindex="-1"></a>plot_abundance </span></code></pre></div>
<p><img src="05-explore_files/figure-html/unnamed-chunk-10-1.png" width="672" /></p>
<p><strong>Density plot</strong> shows the overall abundance distribution for a given
taxonomic group. Let us check the relative abundance of Firmicutes
across the sample collection. The density plot is a smoothened
version of a standard histogram.</p>
<p>The plot shows peak abundances around 30 %.</p>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb66-1"><a href="microbiome-data-exploration.html#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset data by taking only Firmicutes</span></span>
<span id="cb66-2"><a href="microbiome-data-exploration.html#cb66-2" aria-hidden="true" tabindex="-1"></a>tse_firmicutes <span class="ot">&lt;-</span> tse_phylum[<span class="st">&quot;Firmicutes&quot;</span>]</span>
<span id="cb66-3"><a href="microbiome-data-exploration.html#cb66-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-4"><a href="microbiome-data-exploration.html#cb66-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Gets the abundance table</span></span>
<span id="cb66-5"><a href="microbiome-data-exploration.html#cb66-5" aria-hidden="true" tabindex="-1"></a>abundance_firmicutes <span class="ot">&lt;-</span> <span class="fu">assay</span>(tse_firmicutes, <span class="st">&quot;relabundance&quot;</span>)</span>
<span id="cb66-6"><a href="microbiome-data-exploration.html#cb66-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-7"><a href="microbiome-data-exploration.html#cb66-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Creates a data frame object, where first column includes abundances</span></span>
<span id="cb66-8"><a href="microbiome-data-exploration.html#cb66-8" aria-hidden="true" tabindex="-1"></a>firmicutes_abund_df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">t</span>(abundance_firmicutes))</span>
<span id="cb66-9"><a href="microbiome-data-exploration.html#cb66-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Rename the first and only column</span></span>
<span id="cb66-10"><a href="microbiome-data-exploration.html#cb66-10" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(firmicutes_abund_df) <span class="ot">&lt;-</span> <span class="st">&quot;abund&quot;</span></span>
<span id="cb66-11"><a href="microbiome-data-exploration.html#cb66-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-12"><a href="microbiome-data-exploration.html#cb66-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Creates a plot. Parameters inside feom_density are optional. With </span></span>
<span id="cb66-13"><a href="microbiome-data-exploration.html#cb66-13" aria-hidden="true" tabindex="-1"></a><span class="co"># geom_density(bw=1000), it is possible to adjust bandwidth.</span></span>
<span id="cb66-14"><a href="microbiome-data-exploration.html#cb66-14" aria-hidden="true" tabindex="-1"></a>firmicutes_abund_plot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(firmicutes_abund_df, <span class="fu">aes</span>(<span class="at">x =</span> abund)) <span class="sc">+</span> </span>
<span id="cb66-15"><a href="microbiome-data-exploration.html#cb66-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">color=</span><span class="st">&quot;darkred&quot;</span>, <span class="at">fill=</span><span class="st">&quot;lightblue&quot;</span>) <span class="sc">+</span> </span>
<span id="cb66-16"><a href="microbiome-data-exploration.html#cb66-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Relative abundance&quot;</span>, <span class="at">title =</span> <span class="st">&quot;Firmicutes&quot;</span>) <span class="sc">+</span></span>
<span id="cb66-17"><a href="microbiome-data-exploration.html#cb66-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span> <span class="co"># Changes the background</span></span>
<span id="cb66-18"><a href="microbiome-data-exploration.html#cb66-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">label =</span> scales<span class="sc">::</span>percent)</span>
<span id="cb66-19"><a href="microbiome-data-exploration.html#cb66-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-20"><a href="microbiome-data-exploration.html#cb66-20" aria-hidden="true" tabindex="-1"></a>firmicutes_abund_plot</span></code></pre></div>
<p><img src="05-explore_files/figure-html/unnamed-chunk-11-1.png" width="672" /></p>
<p>For more visualization options and examples, see the <a href="https://microbiome.github.io/miaViz/articles/miaViz.html">miaViz vignette</a>.</p>
</div>
<div id="exercises-optional" class="section level2 hasAnchor" number="6.3">
<h2><span class="header-section-number">6.3</span> Exercises (optional)<a href="microbiome-data-exploration.html#exercises-optional" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Explore some of the following questions on your own by following
<a href="https://microbiome.github.io/OMA/">online examples</a>. Prepare a
reproducible report (Rmarkdown), and include the code that you use to
import the data and generate the analyses.</p>
<ul>
<li><p><strong>Abundance table</strong> Retrieve the taxonomic abundance table from the
example data set (TSE object). Tip: check “assays” in <a href="https://microbiome.github.io/OMA/data-introduction.html#loading-experimental-microbiome-data">data import
section</a></p></li>
<li><p>How many different samples and genus-level groups this phyloseq
object has? Tips: see dim(), rowData()</p></li>
<li><p>What is the maximum abundance of Akkermansia in this data set? Tip:
aggregate the data to Genus level with agglomerateByRank, pick
abundance assay, and check a given genus (row) in the assay</p></li>
<li><p>Draw a histogram of library sizes (total number of reads per
sample). Tip: Library size section in
<a href="https://microbiome.github.io/OMA/quality-control.html">OMA</a>. You
can use the available function, or count the sum of reads per
sample by using the colSums command applied on the abundance
table. Check <a href="https://www.nature.com/articles/nature24460">Vandeputte et
al. 2017</a> for further
discussion on the differences between absolute and relative
quantification of microbial abundances.</p></li>
<li><p><strong>Taxonomy table</strong> Retrieve the taxonomy table and print out the
first few lines of it with the R command head(). Investigate how
many different phylum-level groups this phyloseq object has? Tips:
rowData, taxonomicRanks in
<a href="https://microbiome.github.io/OMA/taxonomic-information.html#functions-to-access-taxonomic-information">OMA</a>.</p></li>
<li><p><strong>Sample metadata</strong> Retrieve sample metadata. How many patient
groups this data set has? Draw a histogram of sample
diversities. Tips: colData</p></li>
<li><p><strong>Subsetting</strong> Pick a subset of the data object including only
ADHD individuals from Cohort 1. How many there are? Tips: subsetSamples</p></li>
<li><p><strong>Transformations</strong> The data contains read counts. We can convert
these into relative abundances and other formats. Compare abundance
of a given taxonomic group using the example data before and after
the compositionality transformation (with a cross-plot, for
instance). You can also compare the results to CLR-transformed data
(see e.g. <a href="https://www.frontiersin.org/articles/10.3389/fmicb.2017.02224/full">Gloor et
al. 2017</a>)</p></li>
<li><p><strong>Visual exploration</strong> Visualize the population distribution of
abundances for certain taxonomic groups. Do the same for
CLR-transformed abundances. Tip: assays, transformCounts</p></li>
<li><p>Experiment with other data manipulation tools from
<a href="https://microbiome.github.io/OMA/taxonomic-information.html#functions-to-access-taxonomic-information">OMA</a>.</p></li>
<li><p>Example solution: <a href="06-3-ex-sol-ADHD.html">Solutions</a></p></li>
</ul>

<pre><code>## Loading required package: SummarizedExperiment</code></pre>
<pre><code>## Loading required package: MatrixGenerics</code></pre>
<pre><code>## Loading required package: matrixStats</code></pre>
<pre><code>## 
## Attaching package: &#39;MatrixGenerics&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:matrixStats&#39;:
## 
##     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,
##     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,
##     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,
##     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,
##     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,
##     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,
##     colWeightedMeans, colWeightedMedians, colWeightedSds,
##     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,
##     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,
##     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,
##     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,
##     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,
##     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,
##     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,
##     rowWeightedSds, rowWeightedVars</code></pre>
<pre><code>## Loading required package: GenomicRanges</code></pre>
<pre><code>## Loading required package: stats4</code></pre>
<pre><code>## Loading required package: BiocGenerics</code></pre>
<pre><code>## 
## Attaching package: &#39;BiocGenerics&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:stats&#39;:
## 
##     IQR, mad, sd, var, xtabs</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     anyDuplicated, append, as.data.frame, basename, cbind, colnames,
##     dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,
##     grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,
##     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,
##     rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,
##     union, unique, unsplit, which.max, which.min</code></pre>
<pre><code>## Loading required package: S4Vectors</code></pre>
<pre><code>## 
## Attaching package: &#39;S4Vectors&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     expand.grid, I, unname</code></pre>
<pre><code>## Loading required package: IRanges</code></pre>
<pre><code>## Loading required package: GenomeInfoDb</code></pre>
<pre><code>## Loading required package: Biobase</code></pre>
<pre><code>## Welcome to Bioconductor
## 
##     Vignettes contain introductory material; view with
##     &#39;browseVignettes()&#39;. To cite Bioconductor, see
##     &#39;citation(&quot;Biobase&quot;)&#39;, and for packages &#39;citation(&quot;pkgname&quot;)&#39;.</code></pre>
<pre><code>## 
## Attaching package: &#39;Biobase&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:MatrixGenerics&#39;:
## 
##     rowMedians</code></pre>
<pre><code>## The following objects are masked from &#39;package:matrixStats&#39;:
## 
##     anyMissing, rowMedians</code></pre>
<pre><code>## Loading required package: SingleCellExperiment</code></pre>
<pre><code>## Loading required package: TreeSummarizedExperiment</code></pre>
<pre><code>## Loading required package: Biostrings</code></pre>
<pre><code>## Loading required package: XVector</code></pre>
<pre><code>## 
## Attaching package: &#39;Biostrings&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:base&#39;:
## 
##     strsplit</code></pre>
<pre><code>## Loading required package: MultiAssayExperiment</code></pre>
<pre><code>## Loading required package: ggplot2</code></pre>
<pre><code>## Loading required package: ggraph</code></pre>
<pre><code>## 
## Attaching package: &#39;dplyr&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:Biostrings&#39;:
## 
##     collapse, intersect, setdiff, setequal, union</code></pre>
<pre><code>## The following object is masked from &#39;package:XVector&#39;:
## 
##     slice</code></pre>
<pre><code>## The following object is masked from &#39;package:Biobase&#39;:
## 
##     combine</code></pre>
<pre><code>## The following objects are masked from &#39;package:GenomicRanges&#39;:
## 
##     intersect, setdiff, union</code></pre>
<pre><code>## The following object is masked from &#39;package:GenomeInfoDb&#39;:
## 
##     intersect</code></pre>
<pre><code>## The following objects are masked from &#39;package:IRanges&#39;:
## 
##     collapse, desc, intersect, setdiff, slice, union</code></pre>
<pre><code>## The following objects are masked from &#39;package:S4Vectors&#39;:
## 
##     first, intersect, rename, setdiff, setequal, union</code></pre>
<pre><code>## The following objects are masked from &#39;package:BiocGenerics&#39;:
## 
##     combine, intersect, setdiff, union</code></pre>
<pre><code>## The following object is masked from &#39;package:matrixStats&#39;:
## 
##     count</code></pre>
<pre><code>## The following objects are masked from &#39;package:stats&#39;:
## 
##     filter, lag</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     intersect, setdiff, setequal, union</code></pre>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="importing-microbiome-data.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="alpha-diversity.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["radboud2021_material.pdf"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
